### ISSUE TITLE
Low-Rank Subspaces in GANs

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#146]
## TL; DR

- Perform precise local editing via manipulation on latent vectors discovered from low-rank factorization of image/latent Jacobian.

## Motivation

- Semantic image editing methods typically modify images globally, lacking the ability to alter images in spatially controlled manner.
- Null space obtained from low-rank factorization enables spatially disentangled control on image regions.

## Prior Works

- SeFa (Closed Form Factorization..., CVPR'21) : discover major latent directions through linear approximation of generator function.

## Method

![image](https://user-images.githubusercontent.com/46308407/147906700-45aec6cf-213d-4e24-a4a0-8df5b0438337.png)

- Similar to SeFa, want to discover major directions for image manipulation.

![image](https://user-images.githubusercontent.com/46308407/147906735-4fe3b52a-ad77-4f2f-9008-249fa42d3c2a.png)

- This theorem suggests that the rank of Jacobian w.r.t. each layer's activation is progressively reduced, at last remaining only r vectors corresponding to semantic attributes.

![image](https://user-images.githubusercontent.com/46308407/147906820-579c24c8-7c6e-4a20-914d-e2fd990c22d8.png)
![image](https://user-images.githubusercontent.com/46308407/147906833-b2abe8d6-56d1-45fb-b100-628e98541a59.png)
![image](https://user-images.githubusercontent.com/46308407/147906849-4adcd183-9c3e-47b9-9511-639888756cf1.png)

- Numerically (and empirically) the Jacobian is usually full-rank due to non-linear operations on noises.
- Hence, admit noise input, and still approximate the Jacobian with low-rank factorization *L* (where * norm represents the nuclear norm)

![image](https://user-images.githubusercontent.com/46308407/147906941-bf72865e-edcf-4e6d-a8c2-8b7980bd3617.png)

- Provide a mask indicating the region to be modified, and acquire major latent vectors corresponding to largest singular values.
- For the remaining non-masked regions, similarly obtain low-rank factorization of Jacobian (w.r.t. the region) and keep the ***null space***.
- Vectors that have large impact on masked region while simultaneously belonging to the null space for the remaining region is the latent direction that yields spatially disentangled control.
- Hence, for major latent vectors in region A (masked), project them to the null space of Jacobian for region B (unmasked).

![image](https://user-images.githubusercontent.com/46308407/147907033-0c02b86e-a9b3-4957-bb0d-513bbf19d77c.png)

- As sometimes it can be impossible to alter certain region without affecting the others (e.g., opening mouth with chin kept unchanged), relax the condition. (instead of r non-zero singular values, keep k < r singular values of B for null space projection).

### Baseline and Dataset

- StyleGAN2: FFHQ, LSUN
- BigGAN: ImageNet

![image](https://user-images.githubusercontent.com/46308407/147907151-d1c9adda-6e0c-43e6-b0ae-ee4b0b4d5f0c.png)
![image](https://user-images.githubusercontent.com/46308407/147907175-77a542c3-1a90-4940-9edf-b23e61f630b3.png)
![image](https://user-images.githubusercontent.com/46308407/147907180-e25a2832-7aa3-42bd-979b-7b6aecd0b0bd.png)
![image](https://user-images.githubusercontent.com/46308407/147907207-5f400fd1-d07a-4279-b04e-c521f0169408.png)

## Comments

- Linear algebra papers are always welcome. So insightful yet elegant in their simplicity. (similar impression to SeFa)
- The idea of obtaining Jacobian from a masked image region, and projecting the major latent vectors to the null space of Jacobian of remaining region is very straight-forward and reasonable. 
- The fact that it can be applied to different baseline models is also impressive.
- As we can see from the mask-robustness experiments, (and other studies on semantic segmentation using GANs), it is true that well-trained GAN models are segmentation / semantic - aware, easy to be employed with relatively coarse and rough control input. 
- I believe this characteristic is more clear with StyleGAN2, as it is well-known for its well-aligned semantics. The figure is also (presumably) from StyleGAN2. (FFHQ, LSUN-church, LSUN-car)
